README.rst
rtcr
====

Recover T Cell Receptor (RTCR) is a pipeline for complete and accurate recovery
of TCR repertoires from high throughput sequencing data.

Installation
============

RTCR requires an external aligner, we have chosen for Bowtie 2, but if you
would like to use your own, read the section "switching aligners". The pipeline
has been successfully tested with python version 2.7.3.

1. Install `Bowtie 2 <http://bowtie-bio.sourceforge.net/bowtie2/index.shtml>`_

2. Install the `RTCR package <http://uubram.github.io/RTCR>`_

   Download the zip file and extract it. Bowtie 2 is not included in this file
   and needs to be downloaded and installed separately (see step 1). Open a
   terminal and go to the newly extracted RTCR folder and run::

        python setup.py install
   
   To test if everything went well, type "rtcr" (and press enter). This should
   result in a help being printed showing the command line options of RTCR.

3. Configure RTCR
   
   RTCR needs to know where Bowtie 2 is installed, specifically it needs to know
   the directory where the 'bowtie2-build' and 'bowtie2' executables reside::

        rtcr Config Aligner.location=/path/to/bowtie2/
   
   For more details on configuring RTCR, see the "Configuration" section

Known installation issues
=========================

If there are permission issues (e.g. not having root access) run::
        
        python setup.py install --user

If the vtrie module fails to install, first run::

        pip install vtrie

Quick start
===========

Analysing the non-barcoded sample::

        mkdir my_analysis
        cd my_analysis
        rtcr run -i ../reads.fq.gz

Analysing the barcoded sample::

        mkdir my_barcoded_analysis
        cd my_barcoded_analysis
        rtcr Checkout -rc -b ../barcodes.txt -i ../umi_reads.fq.gz
        rtcr umi_group_ec -i S1.fastq -o S1_umi_group_ec.fastq
        rtcr run -i S1_umi_group_ec.fastq

Usage
=====

RTCR has a single entry-point, the "rtcr" command, with which everything is
run. With different arguments, RTCR can be configured, instructed to do barcode
error correction, or run an analysis.

RTCR takes FastaQ files as input. These files can be zipped to save disk space.
As an example, let us assume you have a zipped dataset called "reads.fastq.gz"
in a directory "/data", and the reads are from a human TcR-beta repertoire.

First, create an empty directory, for example::

        cd /data
        mkdir my_analysis
        cd my_analysis

Now we can analyse the dataset as follows::

        rtcr run --reads ../reads.fastq.gz

RTCR will produce several files (see "Output files" section for more details),
but the retrieved clones will end up in a tab-delimited file called
"results.tsv". If something went wrong, please check the "rtcr.log" file in the
same directory.

.. IMPORTANT::

        Start analyses in an empty folder. If the alignments file
        "alignments.sam.gz" already exists, RTCR will *not* perform alignments
        and use this file instead.

.. NOTE::

        If you have a paired-end HTS dataset, please merge the read-pairs first
        using a program such as pear_.

To use a different germline reference, for example to analyse a mouse TcR-alpha
repertoire HTS dataset (e.g. "mouse.fastq"), use the "--species" and "--gene"
options as follows::

        rtcr run --reads mouse.fastq --species MusMusculus --gene TRA

Finally, if you want to prevent RTCR from merging clones with the same CDR3 but
different VJ combination, use the "--no_VJ_collapsing" option.

Output files
============

The "rtcr run" command produces many files, each of these will be explained
below:

- alignments.sam.gz
        SAM records from the aligner, with alignments between the
        germline V and J sequences and the reads. The VJ alignments to the
        reads occur together, starting with the V alignments. Hence, the first
        record is a V alignment, the second a J alignment to the same read.

- aln_stats.csv
        Alignments statistics for the different sequence lengths
        (CDR3 by default; "seqlen" column), counting the number of bases ("n"),
        mismatches ("mm"), insertions ("ins"), and deletions ("dels")
        encountered in the alignments of the V or J ("region" column). This
        data is used to calculate the various error rates that RTCR uses.

- Q_mm_stats.csv
        For the V and J region alignments ("region" column), the
        number of bases ("n"), and number of mismatches ("mm") found of
        the bases having a certain Phred score ("Q"). This data is used by
        RTCR to calculate a normalization factor.

- Q_mm_stats_plot.pdf
        Graphical representation of Q_mm_stats.csv. This file is not created
        if matplotlib is not installed.

- r.dat
        Raw clones identified by RTCR before error correction.

- rqi.dat
        Clones after QMerge and IMerge algorithms have run.

- rqil.dat
        Clones after QMerge, IMerge, and LMerge algorithms have run.

- rqiln.dat
        Clones after QMerge, IMerge, LMerge, and NMerge algorithms have run.

- results.tsv
        Final list of clones after error correction and post-processing.

.. NOTE::

        Clones produced during the run are output to .dat files (see above).
        These can be converted to the same format as results.tsv using the
        Convert option. For example::
        
                rtcr Convert -i r.dat -o r.tsv

Analysing a barcoded HTS dataset
================================

First RTCR needs to identify the unique molecular identifiers (UMIs) in the
reads. For this it requires the sequence of the primer(s) that contain the UMI.
Let's assume we have a barcoded HTS dataset ("umi_reads.fastq") with one sample
and a 12bp long UMI.

For this create a file call "barcodes.txt" with the following contents::

        S1      aagcagtggtaTCAACGcagagNNNNNNNNNNNNcttggggg

In the first column is the name of the sample, here "S1". The second column,
separated from the first by a tab, contains the primer sequence where the "N"
bases denote the location of the UMI. To look for the UMI, RTCR will search the
read for a 'seed' sequence, that is indicated in the primer by a stretch of
upper case bases (non-N). In the above it is "TCAACG". This seed sequence is
required as RTCR will search the read for a perfect match to this sequence, and
then search for the remaining lower-case DNA bases. By default there are only
2 mismatches allowed in the lower-case bases. To ask RTCR to search for the
UMIs, run::

        rtcr Checkout -f reads.fastq -b barcodes.txt -rc

The "-rc" switch is used to tell RTCR to also look for the UMI in the reverse
complement of the reads. The above should create a file called "S1.fastq". This
file contains the reads in which RTCR managed to identify a UMI.

Next, to perform barcode error correction::

        rtcr umi_group_ec -i S1.fastq -o S1_umi_group_ec.fastq

The "S1_umi_group_ec.fastq" file contains the barcode error corrected reads.
After this one can perform the regular HTS analysis the same as for
non-barcoded HTS datasets::

        rtcr run --reads S1_umi_group_ec.fastq

Analysing a paired-end barcoded HTS dataset
===========================================

We here assume the raw reads have been merged using a program such as pear_. In
the latter case, we can expect to have the following files::

        reads.assembled.fastq
        reads.unassembled.forward.fastq
        reads.unassembled.reverse.fastq
        reads.discarded.fastq

If almost all reads were successfully assembled, it is possible to continue
with only the reads.assembled.fastq file::

        rtcr Checkout -p reads.assembled.fastq -rc -b barcode.txt

However, if many reads were not assembled, then it is possible to take along
the unassembled reads as follows::

        rtcr Checkout -f reads.unassembled.forward.fastq -r reads.unassembled.reverse.fastq -rc -b barcode.txt

The output of the above command depends on barcode.txt, which contains the
barcode(s) rtcr should look for in the reads and sample names. If there is one
sample called "S1", then the above Checkout commands produce the following
three files (in order of the commands)::

        S1_R12.fastq
        S1_R1.fastq
        S1_R2.fastq

These files contain all the read pairs (assembled (R12) and unassembled
(R1 and R2)) in which rtcr was able to find a barcode.

.. NOTE::

        Names of reads in the above fastq files will have the UMI appended to
        the name, "UMI:XXX:YYY:ZZZ", where XXX is the read source (e.g. "R1"),
        and YYY and ZZZ are the UMI nt sequence and ascii encoded (Phred+33)
        base quality scores.

Next, use umi_group_ec for barcode error correction::

        cat S1_R1.fastq S1_R2.fastq S1_R12.fastq | rtcr umi_group_ec -o S1_umi_group_ec.fastq

.. NOTE::

        1. 
                   Currently, there is no check if a read-pair contains a CDR3 in both
                   R1 and R2. Therefore, it is technically possible for an
                   unassembled read to provide a CDR3 twice (though then the
                   question is why assembly failed).
        2. 
                   The format of the read names in S1_umi_group_ec.fastq is
                   "UMI:XXX:YYY:ZZZ:WWW", where XXX is the source e.g. "R1",
                   YYY is the UMI nt sequence, ZZZ is a fraction where the
                   numerator indicates the UMI group number and the denominator
                   the number of UMI groups that share the same UMI, and WWW is
                   the number of reads in the current UMI group.

Finally, rtcr can be run on the barcode corrected reads::

        rtcr run --reads S1_umi_group_ec.fastq

Configuration
=============

RTCR can be easily configured using the "rtcr Config" command. To find out its
usage, type::

        rtcr Config -h

To see the entire configuration file (can be big), type::

        rtcr Config

If you'd rather edit the config file directly, search for the "config.ini" file
in the "RTCR" folder of the package.

The ini file contains different sections, denoted with the brackets "[" and
"]". These sections contain the different settings of RTCR. To check the value
of a setting, say for example the default germline reference gene, type the
name of its corresponding section and name of the key (in this case it is
the key "gene" in section "Defaults")::

        rtcr Config Defaults.gene

The above should print out "TRB". Let's for example change this default to
TCR alpha chains::

        rtcr Config Defaults.gene=TRA

Switching aligners
==================

To run rtcr with a different aligner, the values in the "[Aligner]" section
of its configuration file (see "Configuration") should be updated. There are
several requirements for the new aligner:
1) It should support receiving FastaQ records via stdin (standard in)
2) It should support writing SAM file format output to stdout (standard out)

If the aligner can do those things, then tell RTCR where the aligner is
located with the "location" key. If the aligner does not build an
index, empty out the corresponding keys as follows::

        rtcr Config Aligner.cmd_build_index=
        rtcr Config Aligner.args_build_index=

The "args_XXX" keys are arguments that RTCR passes to commands given in the
"cmd_XXX" keys. Before alignment, RTCR produces a reference fasta file from the
germline reference and passes this (using the "args_build_index" key) to the
command to build an index ("cmd_build_index" key).  The "%(ref_fn)s" and
"%(index_fn)s" in the "args_build_index" key refer to the reference fasta file
and index filenames that RTCR uses internally.

RTCR first aligns the V genes to the reads and then the J genes. It is possible
to run the aligner with different arguments for both with the "args_align_v"
and "args_align_j" keys respectively. Any arguments that are the same for both
can be put in "args_align_base". The "%phred_encoding)s" and "%(n_threads)s"
parts of the arguments will contain the phred encoding (33 or 64) and number
of threads respectively. It is optional to use these in the arguments.

Analyzing RTCR output with tcR R package
========================================

Data analysis of immune receptor repertoires can be performed using tcR_, a
package for the R_ software environment. RTCR provides an R script,
named 'tcR_RTCR_parser.R', for loading RTCR output into the tcR package for
analysis. See below for an example of how to load RTCR output from inside R::

        source("tcR_RTCR_parser.R")
        rtcr_data <- parse.rtcr("results.tsv")

.. _pear: http://sco.h-its.org/exelixis/web/software/pear/
.. _tcR: https://cran.r-project.org/web/packages/tcR/index.html
.. _R: https://www.r-project.org/


changelog
# Change Log
## [0.5.2] 2023-09-10
### Fixed
- setup.py, running 'python setup.py develop' no longer results in
"ValueError: 'editdistance/bycython.pyx' doesn't match any files" by setting
editdistance version to be 0.3.1 (version 0.6.2 gave the ValueError).
- fileio.py, IMGTAlleleContainerFormat can read IMGT reference fasta.
- convert.py, removed unreachable code
## [0.5.1] 2021-02-28
### Added
- umi_group_ec.py and checkout.py show default values for their arguments,
including when they are called using 'rtcr umi_group_ec -h' or
'rtcr Checkout -h'
## [0.5.0] 2020-09-04
### Changed
- pipeline.py and convert.py, output clones in AIRR format
### Added
- config.ini, 'output_not_ok_fn' option specifying the filename for clones
that are discarded by the post-processing after error correction
- docs/README.md, "results.tsv" section explaining the new output format,
including custom fields that are not specified by the AIRR format.
"Analysing RTCR output with immunarch R package" section.
### Removed
- config.ini, 'include_cysphe', 'output_hdr', and 'output_fmt' options
- tcR_RTCR_parser.R, because it cannot parse RTCR output which is now in
AIRR format
- docs/README.md, "Analyzing RTCR output with tcR R package" section.
### Fixed
- docs/README.md, barcode example "Quick start" section works with current
version of RTCR
## [0.4.6]
### Fixed
- checkout.py, when searching both orientations for master also both
orientations are searched for the slave.
- checkout.py, slave UMI is properly retrieved when the slave is found on the
reverse complement of a read.
## [0.4.5]
### Added
- barcode.py, IUPAC DNA codes allowed in non-seed part of the adapter.
### Fixed
- checkout.py, reported number of records processed matches actual number of
records processed.
- checkout.py, handle adapters that have the same master but different slave
sequences.
## [0.4.4]
### Fixed
- setup.py, drop reference to non-existant README.rst, include docs/README.md
## [0.4.3]
### Added
- known issue to README.rst about the pipeline exiting with a
"Error in worker pool" message in case of low memory (on linux the result of
the OOM killer taking out a child process of the pipeline).
- docs/_config.yml for loading github-pages minimal theme and add download
buttons to the page.
### Changed
- Converted README.rst to README.md and placed it in /docs folder for use with
github-pages. Thanks to Gijs Schroder (user seigenrocks on github) for helping
to get github pages working again for RTCR.
### Fixed
- pipeline exits when child process dies instead of waiting indefinitely for
results (same 'fix' in version 0.4.0 made pipeline report a child died, but
neglected to make the pipeline exit)
- "--debug" option of "rtcr run" sets debug level on root logger so that
loggers in the other modules are also using the debug level
- "--verbose" option of "rtcr run" command adds StreamHandler to root logger
so that loggers in the other modules are also using this handler.
## [0.4.2]
### Fixed
- Corrected example command to "Checkout" unassembled reads.
## [0.4.1]
### Fixed
- README.rst, indentation of enumeration in installation section. Removed
superfluous bowtie2 reference at the end. Replaced RTCR hyperlink target at
the end with an in place "RTCR package" hyperlink (fixes "Duplicate implicit
target name" message from rstcheck program).
## [0.4.0]
### Added
- filter to umi_group_ec. Reads that are too dissimilar to an existing
consensus sequence with the same UMI, are discarded if their score (from
get_offset method) is less than "min_score4merge".
- progress indicator for umi_group_ec.
- progress indicator for QMerge&IMerge, and NMerge
- "output_hdr" and "output_fmt" options to config.ini for customization of the
way clones are output to text (currently ".tsv") files.
- "rtcr Convert" option. Converts pickled clonesets (.dat files) to text
(.tsv files).
### Fixed
- Pipeline exits when child process dies (instead of waiting indefinitely for
results).
- config.ini, by default thread name is included in log messages
- test_RTCR.py tests the run_umi_group_ec method.
- UMI record format is now "UMI:XXX:YYY:ZZZ" where XXX is the qualifier (e.g.
"R1"), and YYY and ZZZ are the UMI nt sequence and ascii encoded (Phred+33)
base quality scores. This fixes an issue as the base quality string (here the
ZZZ part) contains the delimiter (":", i.e. Phred 25).
- align.py minimally assigns one thread to the V and J alignment processes, so
running RTCR with number of threads set to one ('-t 1' option) no longer
causes a broken pipe ("IOError: [Errno 32] Broken pipe"), which happened
because the J-alignment process would be given 0 threads to work with in
align.py.
- trie.py now removes nodes in Trie.__delitem__ function  without throwing
AttributeError when trying to access the node.isset attribute (added
Node.clear() function which itself resets the Node._isset attribute).
- Update author email in setup.py to gmail account
- Requirement of python 2.7+ in setup.py (python_require option)
## [0.3.2]
### Added
- Change log
- command line arguments (sys.argv) in pipeline log output (rtcr.log)
### Changed
- Checkout can handle three file inputs with forward, reverse, or merged
reads. Output will be "_R1.fastq" and "_R2.fastq" for master and slave
containing reads in the case of forward and reverse fastq files. Output will be
"_R12.fastq" for merged read pairs.
- Checkout adds a qualifier ("R1", "R2", or "R12") to each UMI
- umi_group_ec handles UMI qualifiers for separating out forward, reverse,
and merged read pairs.
### Fixed
- Checkout now reports "100%" when it has finished processing all
the reads.
- Pipeline responds to CTRL-C
- Suppresses 'No handlers could be found for logger' message when using rtcr
libraries without configuring a logger.
- build_clone method in clone.py no longer accepts 0 length junction nt seqs,
this prevents "ValueError: len(seq) < 1" from ConsensusQSequence (in cseq.c)

rtcr/align.py
#!/usr/bin/env python

import logging
logger = logging.getLogger(__name__)
logger.addHandler(logging.NullHandler())

from subprocess import check_call, Popen, PIPE, STDOUT
import os
import threading as th
from Queue import Queue
import shlex
import shutil

from fileio import SAMRecord, string2SAMRecord, FastaFormat
from util import TemporaryDirectory
from seq import cigar_intervals

try:
    from subprocess import DEVNULL # Py3k
except ImportError:
    import os
    DEVNULL = open(os.devnull, "wb")

def _copyfileobj(src, dest):
    with src, dest:
        shutil.copyfileobj(src, dest)

def start_aligner(dirname, ref, region, cmd_build_index,
        args_build_index, cmd_align, args_align, phred_encoding, n_threads):
    """Starts external aligner.

    :dirname: name of directory to put temporary files such as reference
    sequences
    :ref: AlleleContainer-like containing reference sequences
    :region: string, "V" or "J" region
    :cmd_build_index: command that will be run to build an index of the
    reference sequences. If empty "" or None, this command will not be run.
    :args_build_index: string, arguments that will be provided to
    cmd_build_index
    :cmd_align: string, command that will be run to start the aligner
    :args_align: string, arguments that will be provided to cmd_align
    """
    assert region == "V" or region == "J"
    ref_fn = os.path.join(dirname, "%s_ref.fa"%region)
    index_fn = os.path.join(dirname, "%s_index"%region)
    FastaFormat.records_out(open(ref_fn, 'w'),
            ref.get_alleles(region = region))
    cmd = [cmd_build_index] + shlex.split(args_build_index%locals())
    logger.info("Running command: \"%s\""%" ".join(cmd))
    check_call(cmd, stdout = DEVNULL, stderr = STDOUT)
    cmd = [cmd_align] + shlex.split(args_align%locals())
    logger.info("Running command: \"%s\""%" ".join(cmd))
    return Popen(cmd, bufsize = -1, executable = None, stdin = PIPE,
            stdout = PIPE, stderr = DEVNULL) 

def trim_and_send(v_aligner, j_aligner, results):
    """Trims V-region from reads and sends the remainder to the aligner for
    J-region alignment.

    :v_aligner: Popen object, interface to aligner process for V alignments
    :j_aligner: Popen object, interface to aligner process for J alignments
    :results: Queue, any V alignments will be turned into SAMRecord objects and
    put on this Queue.
    """
    # since the V and J records can arrive out of order, attach a
    # unique identifier.
    uid = 0
    with j_aligner.stdin as j_in:
        for line in iter(v_aligner.stdout.readline, ''):
            rec = string2SAMRecord(line)
            if rec is None:
                continue
            results.put((uid, rec))
            extended_qname = '%s:%s'%(uid, rec.QNAME)
            if rec.FLAG & 4:
                trimmed_read = (extended_qname, "", "")
            else:
                qas, qae, tas, tae = cigar_intervals(rec.CIGAR, rec.POS - 1)
                trimmed_read = (extended_qname, rec.SEQ[qae:],
                        rec.QUAL[qae:])
            j_in.write("@%s\n%s\n+\n%s\n"%trimmed_read)
            uid += 1 # Next unique identifier

def collect_vj_output(v_records, j_aligner, results):
    """Collects V and J alignments and stores the corresponding SAMRecords in
    2-tuples in a Queue.

    Since the V and J aligner processes are running simultaneously, the results
    of the alignments are likely to arrive out of order. This method collects
    the results as they arrive and as soon as V and J alignments to the same
    read are found, their results are put onto the results queue.

    :v_records: Queue, providing SAMRecord objects corresponding to V
    alignments.
    :j_aligner: Popen object, interface to aligner that performs J alignments.
    :results: Queue, will receive 2-tuples (v_rec, j_rec), where v_rec is the
    SAMRecord corresponding to the V alignment and j_rec is the SAMRecord to
    the J alignment.
    """
    tmp_v_recs = {} # temporarily unmatched V SAM records
    tmp_j_recs = {} # temporarily unmatched J SAM records
    for line in iter(j_aligner.stdout.readline, ''):
        j_rec = string2SAMRecord(line)
        if j_rec is None:
            continue

        # Retrieve unique identifier from the J record
        j_qname_fields = j_rec.QNAME.split(':')
        j_uid = int(j_qname_fields[0])

        # Remove the uid from the QNAME field
        line_fields = line.split('\t')
        line_fields[0] = ':'.join(j_qname_fields[1:])
        j_rec = string2SAMRecord('\t'.join(line_fields))

        tmp_j_recs[j_uid] = j_rec

        (v_uid, v_rec) = v_records.get()
        tmp_v_recs[v_uid] = v_rec

        while len(tmp_v_recs) > 0:
            key = next(tmp_v_recs.iterkeys())
            if key in tmp_j_recs:
                results.put( (tmp_v_recs.pop(key), tmp_j_recs.pop(key)) )
            else:
                break

def get_vj_alignments(ref, reads, cmd_build_index, args_build_index,
        cmd_align, args_align_v, args_align_j, phred_encoding, n_threads):
    """Align V and J germline reference sequences to reads (in FastQ format).
    Yields (v_rec, j_rec) as 2-tuples, where v_rec is a SAMRecord object
    containing an alignment of a V sequence to a read identified by
    v_rec.QNAME. j_rec is similar, but contains an alignment of a J sequence to
    the trimmed version of the same read.

    :ref: AlleleContainer containing both V and J germline reference alleles.
    :reads: handle to FastQ file 
    :cmd_build_index: command that will be run to build an index of the
    reference sequences. If empty "" or None, this command will not be run.
    :args_build_index: string, arguments that will be provided to
    cmd_build_index
    :cmd_align: string, command that will be run to start the aligner
    :args_align_v: string, arguments that will be provided to cmd_align to
    align V sequences to the reads.
    :args_align_j: string, arguments that will be provided to cmd_align to
    align J sequences to the reads.
    :phred_encoding: string passed to the aligner to tell it which encoding to
    use. (e.g. for bowtie2 "33" or "64" is used).
    :n_threads: maximum number of threads/processes the aligner is allowed to
    use.
    """
    with TemporaryDirectory() as dirname:
        logger.info("Created temporary directory: \"%s\""%dirname)
        # e.g. if n_threads is 7, 4 will be given to v aligner and 3 to the
        # j aligner. This is because the v aligner is first and has to do more
        # work.
        n_threads_v = max(int(round(n_threads / float(2))), 1)
        n_threads_j = max(n_threads // 2, 1)
        v_aligner = start_aligner(dirname, ref, "V", cmd_build_index,
            args_build_index, cmd_align, args_align_v, phred_encoding,
            n_threads_v)
        j_aligner = start_aligner(dirname, ref, "J", cmd_build_index,
            args_build_index, cmd_align, args_align_j, phred_encoding,
            n_threads_j)

        v_aligner_input_thread = th.Thread(name = "input to v_aligner",
                target = _copyfileobj,
                args = (reads, v_aligner.stdin))
        v_aligner_input_thread.start()

        # Trim V-region from reads and send trimmed reads to aligner, for J
        # segment alignment, in separate thread
        v_records = Queue()
        j_aligner_input_thread = th.Thread(name = "input to j_aligner",
                target = trim_and_send,
                args = (v_aligner, j_aligner, v_records))
        j_aligner_input_thread.start()

        vj_records = Queue()
        vj_output_thread = th.Thread(name = "collect vj output",
                target = collect_vj_output,
                args = (v_records, j_aligner, vj_records))
        vj_output_thread.start()

        while vj_output_thread.is_alive() or not vj_records.empty():
            while not vj_records.empty():
                yield vj_records.get()

def main():
    from rtcr_io import read_reference, zopen
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument("-r","--ref", required = True,
            help = "Immune receptor reference file")
    parser.add_argument("--species",
            help = "Which species should be used as reference")
    parser.add_argument("--gene",
            help = "Which gene should be used as reference")
    parser.add_argument("-i","--reads", required = True,
            help = "Fastq file with reads")
    parser.add_argument("-p","--phred_encoding", type=int, required = True,
            help = "Phred encoding (ascii code for a PHRED score of 0)")
    parser.add_argument("--threads", type = int, default = 1,
            help = "Number of threads the aligner should use")
    args = parser.parse_args()

    cmd_build_index = "bowtie2-build"
    args_build_index = r"%(ref_fn)s %(index_fn)s"

    cmd_align = "bowtie2"
    args_align_v = r"-D 20 -R 3 -N 0 -i S,1,0.50 -L 8 " + \
        r"--local -x %(index_fn)s - --phred%(phred_encoding)s " + \
        r"--threads %(n_threads)s"
    args_align_j = r"-D 20 -R 3 -N 0 -i S,1,0.50 -L 8 " + \
        r"--local -x %(index_fn)s - --phred%(phred_encoding)s " + \
        r"--threads %(n_threads)s"

    ref_fn = os.path.realpath(args.ref)
    species = args.species
    gene = args.gene
    reads_fn = os.path.realpath(args.reads)
    phred_encoding = args.phred_encoding
    n_threads = args.threads

    ref = read_reference(ref_fn).get_slice(species = species,gene = gene)
    for v_rec, j_rec in get_vj_alignments(ref,
            zopen(reads_fn, "rb"), cmd_build_index,
            args_build_index, cmd_align, args_align_v, args_align_j,
            phred_encoding, n_threads):
        print "v_rec = %s"%map(str, v_rec)
        print "j_rec = %s"%map(str, j_rec)
    
if __name__ == "__main__":
    main()
